MassSpectrometry/MsDataFile.cs

```diff
@@ -83,63 +83,175 @@ public static int TopNpeakHelper(ref double[] intensities, ref double[] mArray,
+            {
+                mzRangeInOneWindow = mzRangeInOneWindow / (double)filteringParams.NumberOfWindows;
+
+                for (int i = 0; i < (int)filteringParams.NumberOfWindows; i++)
+                {
+                    if (i == 0) // first
+                    {
+                        ranges.Add(scanMin - 0.000000001, (scanMin + mzRangeInOneWindow));
+                    }
+                    else if (i == ((int)filteringParams.NumberOfWindows - 1))//last
+                    {
+                        ranges.Add(scanMin, (scanMin + mzRangeInOneWindow) + 0.000000001);
+                    }
+                    else//middle
+                    {
+                        ranges.Add(scanMin, (scanMin + mzRangeInOneWindow));
+                    }
+                    scanMin += mzRangeInOneWindow;
+                }
+            }
+            else
             {
-                windowSize = 1;
+                ranges.Add(scanRangeMinMz - 0.000000001, scanRangeMaxMz + 0.000000001);
             }
 
-            int windowPeakIndexMinimum = 0;
-            int windowPeakIndexMaximum = windowSize - 1;
+            foreach (Tuple<double, double> range in ranges)
+            {
+                var Temp = new Chemistry.ClassExtensions.TupleList<double, double>();
+                foreach (Tuple<double, double> mzIntensityPair in mzIntensites)
+                {
+                    if (mzIntensityPair.Item1 > range.Item1 && mzIntensityPair.Item1 <= range.Item2)
+                    {
+                        Temp.Add(mzIntensityPair);
+                    }
+                }
+                if (Temp.Count > 0)
+                {
+                    Temp.Sort((x, y) => y.Item2.CompareTo(x.Item2)); //sort tuple in place decending by item 2, reverse by changing x and y
+                    double maxIntensity = Temp.Select(pair => pair.Item2).ToList().Max();
+                    for (int i = 0; i < Math.Min(Temp.Count, filteringParams.NumberOfPeaksToKeepPerWindow.Value); i++)
+                    {
+                        if (WindowMaxNormalizationToValue.HasValue)
+                        {
+                            double normalizedIntensity = Temp[i].Item2 / maxIntensity * (double)WindowMaxNormalizationToValue.Value;
+                            mzIntensites_reduced.Add(new Tuple<double, double>(Temp[i].Item1, normalizedIntensity));
+                        }
+                        else
+                        {
+                            mzIntensites_reduced.Add(Temp[i]);
+                        }
+                    }
+                }
+            }
+
+            // convert merged results to array and sort by m/z
+            mArray = mzIntensites_reduced.Select(i => i.Item1).ToArray();
+            intensities = mzIntensites_reduced.Select(i => i.Item2).ToArray();
+
+            Array.Sort(mArray, intensities);
```

Comment:
Sort before splitting the Peaks into arrays (that way it's only sorting "one" list instead of two)

Reply:
done